src/main/java/BobCat/executor/parser/CommandParser.java

```diff
@@ -0,0 +1,9 @@
+package BobCat.executor.parser;
+
+/**
+ * Interface that all command parsers in BobCat.BobCat must obey. A <code>CommandParser</code> checks if the given command
+ * and commandArgs are appropriate for each other. Specifically, it checks for the number of given arguments.
+ */
+public interface CommandParser {
+    String[] parse(String command, String[] commandArgs);
+}
```

Comment:
Interesting design decision here. May I know why did you decide to create so many different CommandParsers?

Reply:
Hi. The decision was made due to the intention to categorise a *primitive* Command into 3 kinds: a BasicCommand which does read-only operations, a TaskCreation command which creates a new Task, as well as a TaskMutationCommand, which mutates a given Task.

The reason behind this hierarchy is to force an implementer to create commands with a clear, small job i.e. either read-only, create or mutate.  Other kinds of commands can then be created as a composition of these primitive commands, thus forcing reuse of the already written primitive commands. 

The myriad of parsers are written to implement this idea. This leads to BasicCommandParser, CreationCommandParser, and MutationCommandParser. An alternative approach (that I can think of) is to have a `Command` class, with methods `execute` and `parse`. Not too sure which one is better, if there's even a difference, so currently sticking to this approach. 

QueryParser is a general parser that calls the appropriate CommandParser. It does not implement CommandParser. The same holds for StorageParser, which parses the storage file. This is done to allow flexibility in how storage files are written.